Force menu rebuild when window is closed (#10702)

ooikengsiang opened the pull request:
### Description of Change
Manually resign the key window and force the menu bar rebuild when key window is closed, so that the menu bar in MacOS is shown correctly (if the window is close in the correct order it is created). This changes still doesn't fix menu bar showing incorrectly if user switch window focus in MacOS manually or user close window different than the order it created.

### Issues Fixed
Fixes #10702

ooikengsiang committed:
Force menu rebuild when window is closed (#10702)

Manually resign the key window and force the menu rebuild when key window is closed, so that the menu in MacOS is shown correctly.

rmarinho commented:
/azp run

ooikengsiang commented on src/Core/src/Platform/iOS/MauiUISceneDelegate.cs:
```diff
@@ -32,6 +32,15 @@ public virtual void WillConnect(UIScene scene, UISceneSession session, UISceneCo
+			{
+				// manually resign the key window and rebuild the menu
+				Window.ResignKeyWindow();
+				UIMenuSystem
```
I agree, but I found a limitation on MacCatalyst when research for this problem. There aren't any API we can pull to get the focused window. It may sound crazy but iPad actually doesn't need this since only one window will be visible at a time which is the latest window created (also key window). The only API that can track focused windows I found is from AppKit (is there a way we can access AppKit?). That is why my fix only can play around with keywindow. Let me know if you have better idea.

PureWeen replied:
I tried this and it seems to work on catalyst
https://stackoverflow.com/a/75440759

I think you can check if the activeappearance on the UIWindows's traitcollection is active.

ooikengsiang replied:
Previously I tried this solution and it doesn't return the focused window but activated scene only. But I was testing many different methods that time and I might be wrong, let me try again and get back to here later.

ooikengsiang replied:
The active appearance work with some workaround, see my latest changes, now iOS14+ can rebuild menu whenever a window get focused.

ooikengsiang committed:
Rebuild menu when window changed to activate

Rebuild menu when window change from inactive appearance to active appearance or focused for iOS 14+.

PureWeen commented on src/Core/src/Platform/iOS/MauiUIApplicationDelegate.Menu.cs:
```diff
@@ -24,9 +26,43 @@ public override void BuildMenu(IUIMenuBuilder builder)
-				UIApplication.SharedApplication.GetWindow()?.Handler?.PlatformView as UIWindow;
-
+			UIWindow? window = null;
+			if (OperatingSystem.IsIOSVersionAtLeast(14))
```
```suggestion
			if (OperatingSystem.IsMacCatalystVersionAtLeast(14))
```

PureWeen commented on src/Core/src/Platform/iOS/MauiUIApplicationDelegate.Menu.cs:
```diff
@@ -1,4 +1,6 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
We try to stay away from using LINQ with any core code because it can occasionally lead to some performance issues. Can you rework the code to not use linq?

PureWeen commented on src/Core/src/Platform/iOS/MauiUISceneDelegate.cs:
```diff
@@ -32,6 +32,20 @@ public virtual void WillConnect(UIScene scene, UISceneSession session, UISceneCo
+
+			// for iOS 13 only where active apperance is not supported yet
+			// for iOS 14+, see DidUpdateCoordinateSpace
+			if (OperatingSystem.IsIOSVersionAtLeast(13) && !OperatingSystem.IsIOSVersionAtLeast(14))
```
```suggestion
			if (!OperatingSystem.IsMacCatalystVersionAtLeast(14))
```

PureWeen reviewed (changes requested):
Few comments. This PR is amazing thank you! Apologies on the slow churn. I tested this out with a few different permutations of windows and menus and it worked perfectly!!!

ooikengsiang committed:
Update code base on code review

Update code base on review given and replace LINQ code.

ooikengsiang commented:
> Few comments. This PR is amazing thank you! Apologies on the slow churn. I tested this out with a few different permutations of windows and menus and it worked perfectly!!!

Thanks for the review. I have updated the code base on your review.

PureWeen commented:
/azp run
